File: apps/bff/src/modules/task/ports.ts
import type {
  CreateProjectRequest,
  CreateTaskRequest,
  Project,
  SyncChangesResponse,
  Task,
  TaskPatch,
  TaskRangeResponse,
} from '@tooday/shared';

export interface CreateProjectInput extends CreateProjectRequest {
  userId: string;
}

export interface CreateTaskInput extends CreateTaskRequest {
  userId: string;
}

export interface ListTasksRangeInput {
  userId: string;
  /** 'YYYY-MM-DD' inclusive */
  from: string;
  /** 'YYYY-MM-DD' inclusive */
  to: string;
}

/** 의도 기반 부분 업데이트 — patch에 담긴 필드만 최신 행에 적용한다 (필드 단위 LWW) */
export interface UpdateTaskInput {
  userId: string;
  id: string;
  patch: TaskPatch;
}

/** 델타 동기화 — cursor(마지막으로 본 sync seq) 이후의 변경 */
export interface ListChangesInput {
  userId: string;
  cursor: number;
}

/** 소유자 스코프의 단건 참조 — 조회·삭제·프로젝트 필터에서 재사용 */
export interface TaskRefInput {
  userId: string;
  id: string;
}

export interface ListTasksByProjectInput {
  userId: string;
  projectId: string;
}

/** 프로젝트별 태스크 집계 — 목록 화면의 진행률(완료/전체)에 쓴다 (삭제 제외) */
export interface ProjectTaskCounts {
  projectId: string;
  total: number;
  done: number;
}

export interface ProjectStore {
  listByUser(userId: string): Promise<Project[]>;
  findById(input: { userId: string; id: string }): Promise<Project | null>;
  create(input: CreateProjectInput): Promise<Project>;
}

export interface TaskStore {
  /** 소유자가 아니거나 없는(또는 삭제된) 태스크면 null */
  findById(input: TaskRefInput): Promise<Task | null>;
  /** 프로젝트 상세(보드)용 — 그 프로젝트의 삭제되지 않은 태스크 전부(상태 무관) */
  listByProject(input: ListTasksByProjectInput): Promise<Task[]>;
  /** 프로젝트별 완료/전체 태스크 수 — 프로젝트 없는(project_id null) 태스크는 제외 */
  countsByProject(userId: string): Promise<ProjectTaskCounts[]>;
  create(input: CreateTaskInput): Promise<Task>;
  /** 소유자가 아니거나 없는(또는 삭제된) 태스크면 null */
  update(input: UpdateTaskInput): Promise<Task | null>;
  /** 소프트 삭제 — tombstone으로 델타에 실린다. 대상이 없으면(또는 이미 삭제) false */
  remove(input: TaskRefInput): Promise<boolean>;
}

/**
 * 동기화 읽기 — 데이터와 커서를 **단일 스냅샷**에서 함께 읽는다.
 *
 * 커서는 반환된 행들의 max가 아니라 같은 스냅샷에서 읽은 sync_counters에서
 * 도출한다 (커서 ≤ 데이터 스냅샷 보장). 낱개 조회(listRange·changesSince·
 * syncCursor)를 포트에 노출하지 않는 이유: 호출자가 그것들을 별도 커넥션
 * (=별도 스냅샷)에서 조합하면, 두 읽기 사이에 커밋된 변경을 응답에서 빠뜨린 채
 * 커서만 전진시켜 그 변경이 영구 유실되는 구멍이 생긴다.
 */
export interface SyncReadStore {
  /** 초기 로드 — 주간 창 태스크 + 프로젝트 + 커서 베이스라인을 한 스냅샷에서 */
  range(input: ListTasksRangeInput): Promise<TaskRangeResponse>;
  /** 델타 — 커서 이후의 태스크·프로젝트 변경(tombstone 포함) + 전진한 커서를 한 스냅샷에서 */
  changes(input: ListChangesInput): Promise<SyncChangesResponse>;
}

File: rust/crates/bff/src/modules/task/ports.rs
use async_trait::async_trait;
use tooday_shared::{Project, ProjectChange, ProjectColor, Task, TaskChange, TaskPatch};

use crate::platform::errors::StoreResult;

#[derive(Debug, Clone)]
pub struct CreateProjectInput {
    pub user_id: String,
    pub name: String,
    pub color: ProjectColor,
}

#[derive(Debug, Clone)]
pub struct CreateTaskInput {
    pub user_id: String,
    pub title: String,
    pub project_id: Option<String>,
    pub date: String,
    pub start_at: String,
    pub duration_min: i64,
}

#[derive(Debug, Clone)]
pub struct ListTasksRangeInput {
    pub user_id: String,
    /// 'YYYY-MM-DD' inclusive
    pub from: String,
    /// 'YYYY-MM-DD' inclusive
    pub to: String,
}

/// 의도 기반 부분 업데이트 — patch에 담긴 필드만 최신 행에 적용한다 (필드 단위 LWW)
#[derive(Debug, Clone)]
pub struct UpdateTaskInput {
    pub user_id: String,
    pub id: String,
    pub patch: TaskPatch,
}

/// 델타 동기화 — cursor(마지막으로 본 sync seq) 이후의 변경
#[derive(Debug, Clone)]
pub struct ListChangesInput {
    pub user_id: String,
    pub cursor: i64,
}

/// 소유자 스코프의 단건 참조 — 조회·삭제·프로젝트 필터에서 재사용
#[derive(Debug, Clone)]
pub struct TaskRefInput {
    pub user_id: String,
    pub id: String,
}

#[derive(Debug, Clone)]
pub struct ListTasksByProjectInput {
    pub user_id: String,
    pub project_id: String,
}

/// 프로젝트별 태스크 집계 — 목록 화면의 진행률(완료/전체)에 쓴다 (삭제 제외)
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ProjectTaskCounts {
    pub project_id: String,
    pub total: i64,
    pub done: i64,
}

#[async_trait]
pub trait ProjectStore: Send + Sync {
    async fn list_by_user(&self, user_id: &str) -> StoreResult<Vec<Project>>;
    async fn find_by_id(&self, input: &TaskRefInput) -> StoreResult<Option<Project>>;
    async fn create(&self, input: CreateProjectInput) -> StoreResult<Project>;
    async fn changes_since(&self, input: ListChangesInput) -> StoreResult<Vec<ProjectChange>>;
}

#[async_trait]
pub trait TaskStore: Send + Sync {
    async fn list_range(&self, input: ListTasksRangeInput) -> StoreResult<Vec<Task>>;
    /// 소유자가 아니거나 없는(또는 삭제된) 태스크면 None
    async fn find_by_id(&self, input: &TaskRefInput) -> StoreResult<Option<Task>>;
    /// 프로젝트 상세(보드)용 — 그 프로젝트의 삭제되지 않은 태스크 전부(상태 무관)
    async fn list_by_project(&self, input: ListTasksByProjectInput) -> StoreResult<Vec<Task>>;
    /// 프로젝트별 완료/전체 태스크 수 — 프로젝트 없는(project_id null) 태스크는 제외
    async fn counts_by_project(&self, user_id: &str) -> StoreResult<Vec<ProjectTaskCounts>>;
    async fn create(&self, input: CreateTaskInput) -> StoreResult<Task>;
    /// 소유자가 아니거나 없는(또는 삭제된) 태스크면 None
    async fn update(&self, input: UpdateTaskInput) -> StoreResult<Option<Task>>;
    /// 소프트 삭제 — tombstone으로 델타에 실린다. 대상이 없으면(또는 이미 삭제) false
    async fn remove(&self, input: &TaskRefInput) -> StoreResult<bool>;
    async fn changes_since(&self, input: ListChangesInput) -> StoreResult<Vec<TaskChange>>;
    /// 유저의 현재 sync seq — 클라이언트의 초기 커서
    async fn sync_cursor(&self, user_id: &str) -> StoreResult<i64>;
}
